src/jit/asmjit/asmjit.cc

```diff
@@ -0,0 +1,329 @@
+      cc.mov(getReg(insn.rd()), rd);                                           \
+    }                                                                          \
+    break;                                                                     \
+  }
+
+#define PROT_ASMJIT_LU_OP(OP, DATA_TYPE)                                       \
+  case k##OP: {                                                                \
+    cc.mov(rs1, getReg(insn.rs1()));                                           \
+    cc.add(rs1, insn.imm());                                                   \
+    asmjit::InvokeNode *invoke = nullptr;                                      \
+    cc.invoke(                                                                 \
+        &invoke, reinterpret_cast<size_t>(loadHelper<DATA_TYPE>),              \
+        asmjit::FuncSignature::build<DATA_TYPE, CPUState &, isa::Addr>());     \
+    invoke->setArg(0, state_ptr);                                              \
+    invoke->setArg(1, rs1);                                                    \
+    invoke->setRet(0, rd);                                                     \
+    setDst(insn.rd(), rd);                                                     \
+    break;                                                                     \
+  }
+
+namespace prot::engine {
+namespace {
+
+using JitFunction = void (*)(CPUState &);
+
+class AsmJit : public JitEngine {
+public:
+  AsmJit() = default;
+
+private:
+  bool doJIT(CPUState &state) override {
+    const auto pc = state.getPC();
+    auto found = m_cacheTB.find(pc);
+    if (found != m_cacheTB.end()) {
+      found->second(state);
+      return true;
+    }
+
+    const auto *bbInfo = getBBInfo(pc);
+    if (bbInfo == nullptr) {
+      // No such bb yet
+      return false;
+    }
+
+    auto holder = translate(*bbInfo);
+
+    auto [it, wasNew] = m_cacheTB.try_emplace(pc, std::move(holder));
+    assert(wasNew);
+
+    it->second(state);
+
+    return true;
+  }
+
+  [[nodiscard]] JitFunction translate(const BBInfo &info);
+
+  asmjit::JitRuntime runtime;
+  std::unordered_map<isa::Addr, JitFunction> m_cacheTB;
+};
+
+template <typename T> void storeHelper(CPUState &state, isa::Addr addr, T val) {
+  state.memory->write(addr, val);
+}
+
+template <typename T> T loadHelper(CPUState &state, isa::Addr addr) {
+  return state.memory->read<T>(addr);
+}
+
+void syscallHelper(CPUState &state) { state.emulateSysCall(); }
+
+JitFunction AsmJit::translate(const BBInfo &info) {
+  asmjit::CodeHolder code;
+  code.init(runtime.environment());
+
+  asmjit::x86::Compiler cc(&code);
+  auto signature = asmjit::FuncSignature::build<void, CPUState *>();
+  auto *func_node = cc.addFunc(signature);
+
+  auto state_ptr = cc.newUIntPtr();
+  func_node->setArg(0, state_ptr);
+
+  auto getReg = [&state_ptr](auto regId) {
+    return asmjit::x86::dword_ptr(state_ptr, offsetof(CPUState, regs) +
```

Comment:
Consider returning hardwired zero if `reg_id == 0`. Since asmjit uses optimization passes, it can simplify resulting code a lot

Reply:
Okay, I got what you meant

<img width="1513" height="1149" alt="image" src="https://github.com/user-attachments/assets/d815deb7-fee3-48ac-a9f7-704be6727980" />